src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs

```diff
@@ -17,7 +17,7 @@ public MainPipelineExecutor(IServiceProvider rootBuilder, IPipelineCache pipelin
             this.receivePipeline = receivePipeline;
         }
 
-        public async Task Invoke(MessageContext messageContext)
+        public async Task<MessageProcessingResult> Invoke(MessageContext messageContext)
```

Comment:
What is the benefit of this type? Now we are paying an extra allocation per every message that was handled for a feature that is rarely ever used. So either returns true or false or have some kind of singleton that is returned in the good case at least?

Or it could also be a struct with a default value if necessary

Reply:
The idea was first to use a `bool` but then we talked about being able to evolve this in a non-breaking way but perhaps that is YAGNI? Perhaps a struct is the best tradeoff?

Reply:
I was also wondering if we should make this a `struct` because then we can get rid of all the instances in the tests in favour of `default`. It sounds good to me.

Reply:
yea, will make the change